cmd/prow-job-dispatcher/main.go

```diff
@@ -61,9 +83,248 @@ func (o *options) complete(secrets *sets.String) error {
+func getCloudProvidersForE2ETests(jc *prowconfig.JobConfig) sets.String {
+	cloudProviders := sets.NewString()
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			if ct := getCloudProviderFromEnv(job.Spec); ct != "" {
+				cloudProviders.Insert(ct)
+			}
+		}
+	}
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			if ct := getCloudProviderFromEnv(job.Spec); ct != "" {
+				cloudProviders.Insert(ct)
+			}
+		}
+	}
+	for _, job := range jc.Periodics {
+		if ct := getCloudProviderFromEnv(job.Spec); ct != "" {
+			cloudProviders.Insert(ct)
+		}
+	}
+	return cloudProviders
+}
+
+type clusterVolume struct {
+	// [cloudProvider][cluster]volume
+	clusterVolumeMap map[string]map[string]float64
+	// only needed for stable tests: traverse the above map by sorted key list
+	cloudProviders sets.String
+	mutex          sync.Mutex
+}
+
+// findClusterForJobConfig finds a cluster running on a preferred cloud provider for the jobs in a Prow job config.
+// The chosen cluster will be the one with minimal workload with the given cloud provider.
+// If the cluster provider is empty string, it will choose the one with minimal workload across all cloud providers.
+func (cv *clusterVolume) findClusterForJobConfig(cloudProvider string, jc *prowconfig.JobConfig, path string, config *dispatcher.Config, jobVolumes map[string]float64) string {
+	//no cluster in the build farm is from the targeting cloud provider
+	if _, ok := cv.clusterVolumeMap[cloudProvider]; !ok {
+		cloudProvider = ""
+	}
+	var cluster, rCloudProvider string
+	min := float64(-1)
+	cv.mutex.Lock()
+	for _, cp := range cv.cloudProviders.List() {
+		m := cv.clusterVolumeMap[cp]
+		for c, v := range m {
+			if cloudProvider == "" || cloudProvider == cp {
+				if min < 0 || min > v {
+					min = v
+					cluster = c
+					rCloudProvider = cp
+				}
+			}
+		}
+	}
+
+	for k := range jc.PresubmitsStatic {
+		for _, job := range jc.PresubmitsStatic[k] {
+			cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+		}
+	}
+
+	for k := range jc.PostsubmitsStatic {
+		for _, job := range jc.PostsubmitsStatic[k] {
+			cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+		}
+	}
+	for _, job := range jc.Periodics {
+		cv.addToVolume(rCloudProvider, cluster, job.JobBase, path, config, jobVolumes)
+	}
+
+	cv.mutex.Unlock()
+	return cluster
+}
+
+func (cv *clusterVolume) addToVolume(cloudProvider, cluster string, jobBase prowconfig.JobBase, path string, config *dispatcher.Config, jobVolumes map[string]float64) {
+	determinedCluster, canBeRelocated := config.DetermineClusterForJob(jobBase, path)
+	if cluster == string(determinedCluster) || canBeRelocated {
+		cv.clusterVolumeMap[cloudProvider][cluster] = cv.clusterVolumeMap[cloudProvider][cluster] + jobVolumes[jobBase.Name]
+	} else if determinedCloudProvider, ok := config.IsInBuildFarm(determinedCluster); ok {
```

Comment:
Shouldn't it be an error if `!ok`?

Reply:
Hmm. I think we have different interpretations of a cluster in buildFarm:

- (1): any cluster whose kubeconfig is given to Prow is a cluster in buildFarm.
- (2): (1) + the clusters where we want to distribute workload (the jobs that can be run on any cluster)

I use (2) and I think we will not treat app.ci as a cluster in buildFarm.